### PR TITLE
Add platform to `/image/json` endpoint

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

This adds platform (in the format os/arch) to the `GET /images/json` engine endpoint as an optional field (for backwards compatibility). This is the first part to enabling docker images (or API clients) to see the platform when multiple operating systems are supported (such as LCOW after https://github.com/moby/moby/pull/34859 is merged).

@tonistiigi @dnephin @mlaventure @vieux @stevvooe @johnstep PTAL.

@patricklang FYI

Here's an example of what's returned with this change (note this was adding the change on top of #34859 to have an engine with images for multiple operating systems)

```
[
  {
    "Containers": -1,
    "Created": 1506022363,
    "Id": "sha256:a7a97c6ddc217ee8f1b514f7606795055e03c9de5686d5d408854e3478f4d094",
    "Labels": {},
    "ParentId": "sha256:54511612f1c4d97e93430fc3d5dc2f05dfbe8fb7e6259b7351deeca95eaf2971",
    "Platform": "linux/amd64",
    "RepoDigests": null,
    "RepoTags": [
      "committed:latest"
    ],
    "SharedSize": -1,
    "Size": 3671040,
    "VirtualSize": 3671040
  },
  {
    "Containers": -1,
    "Created": 1506022326,
    "Id": "sha256:470ea9aae35c8792eff3ed8ef8cddab4a8fa9deadf5ac0e9b4cd88826d498011",
    "Labels": null,
    "ParentId": "sha256:a597272f193378c0b9e2476fc80c3d60e096ebf914d96e962614b0e99fdbf64e",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "<none>@<none>"
    ],
    "RepoTags": [
      "<none>:<none>"
    ],
    "SharedSize": -1,
    "Size": 3671040,
    "VirtualSize": 3671040
  },
  {
    "Containers": -1,
    "Created": 1506022316,
    "Id": "sha256:28b0016ce298a201a731a49a906aa65f9a644272b28fbdd3c803fea46901ee93",
    "Labels": null,
    "ParentId": "sha256:aef14fd3dfa7d0ab84e148918bfb528aa104313792dc2955dff0e4f5f32f3cde",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "<none>@<none>"
    ],
    "RepoTags": [
      "<none>:<none>"
    ],
    "SharedSize": -1,
    "Size": 3671040,
    "VirtualSize": 3671040
  },
  {
    "Containers": -1,
    "Created": 1505808826,
    "Id": "sha256:b4e78b89bcf31a38eab47da1b2b36235462e83beebacb7479d725367e76572cf",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "mysql@sha256:0dc3dacb751ef46a6647234abdec2d47400f0dfbe77ab490b02bffdae57846ed"
    ],
    "RepoTags": [
      "mysql:latest"
    ],
    "SharedSize": -1,
    "Size": 531633664,
    "VirtualSize": 531633664
  },
  {
    "Containers": -1,
    "Created": 1505805943,
    "Id": "sha256:28525f9a6e46bd6a033b3a3e5f5f64b28d93698b0a23cdb8d66b43b8d8e6b7c9",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "registry@sha256:0f8fe61fa337b8ef02217702ba979b47a7d68717d4628f31592ebff85915f3ba"
    ],
    "RepoTags": [
      "registry:latest"
    ],
    "SharedSize": -1,
    "Size": 60033536,
    "VirtualSize": 60033536
  },
  {
    "Containers": -1,
    "Created": 1505804593,
    "Id": "sha256:1227c4263c8c59af10d93968ad47e154eedbae2c173633cb4a8ccf49ba28fb19",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "postgres@sha256:144221a1e46af653bdda2bbb8e46ad3dd0c3f61ec0430030581031937f1d589d"
    ],
    "RepoTags": [
      "postgres:latest"
    ],
    "SharedSize": -1,
    "Size": 405280768,
    "VirtualSize": 405280768
  },
  {
    "Containers": -1,
    "Created": 1505792207,
    "Id": "sha256:aaf79d45ddb127b8590118cf2d76e3753dd9b9e23ccc35bc8527d5c7743daea7",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "redis@sha256:8a54dcc711406447b3631a81ef929f500e6836b43e7d61005fa27057882159da"
    ],
    "RepoTags": [
      "redis:latest"
    ],
    "SharedSize": -1,
    "Size": 154930176,
    "VirtualSize": 154930176
  },
  {
    "Containers": -1,
    "Created": 1505783482,
    "Id": "sha256:d2327de7e695e105d1f764c5c291c3305cb1b2168960fab9afd2df2e71343532",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "httpd@sha256:a46bd62f5286321ed19875778246f3afe97df5f9e91f0dd369a6ecfa529dbe81"
    ],
    "RepoTags": [
      "httpd:latest"
    ],
    "SharedSize": -1,
    "Size": 285216256,
    "VirtualSize": 285216256
  },
  {
    "Containers": -1,
    "Created": 1505777500,
    "Id": "sha256:2d696327ab2e15a3354ca258249358e244f3219eef76c653d0e6d3bd87f7830f",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "ubuntu@sha256:31371c117d65387be2640b8254464102c36c4e23d2abe1f6f4667e47716483f1"
    ],
    "RepoTags": [
      "ubuntu:latest"
    ],
    "SharedSize": -1,
    "Size": 176949760,
    "VirtualSize": 176949760
  },
  {
    "Containers": -1,
    "Created": 1505417617,
    "Id": "sha256:049d5d32e9f5885379e765b40016f3b9bfef5512884621fd967ec8e9884328e7",
    "Labels": null,
    "ParentId": "",
    "Platform": "windows/amd64",
    "RepoDigests": null,
    "RepoTags": [
      "microsoft/nanoserver:10.0.16292.0",
      "microsoft/nanoserver:latest"
    ],
    "SharedSize": -1,
    "Size": 203681950,
    "VirtualSize": 203681950
  },
  {
    "Containers": -1,
    "Created": 1505412357,
    "Id": "sha256:de1099630c130d82aaaa009f46d0eb04cb84992c3c2c0378f1d1a6a66fd79275",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "node@sha256:d3fae9d9a3f19feb5b891d12f32a43e16d4974a3fa0948bd9073ccc2e2d2a8c2"
    ],
    "RepoTags": [
      "node:latest"
    ],
    "SharedSize": -1,
    "Size": 973082624,
    "VirtualSize": 973082624
  },
  {
    "Containers": -1,
    "Created": 1505337529,
    "Id": "sha256:39f5c173b5d403baaf377eb0af4ad8b3db1952a18308b55d51fe56a7e67808e1",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "mongo@sha256:bb962ca3964c70069e652e814f7a495db19643ecb5bb62d799f621a68f4da90c"
    ],
    "RepoTags": [
      "mongo:latest"
    ],
    "SharedSize": -1,
    "Size": 438310400,
    "VirtualSize": 438310400
  },
  {
    "Containers": -1,
    "Created": 1505325809,
    "Id": "sha256:2569518fadd0d4c4f25eca2fd6b3efc880dfee2d75b983fe58f81f3dcfddee69",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "swarm@sha256:cc94ce2e6d4603fe9077919db5a4508f7441cad7ae57ccd420fd14af3b3d4383"
    ],
    "RepoTags": [
      "swarm:latest"
    ],
    "SharedSize": -1,
    "Size": 18089472,
    "VirtualSize": 18089472
  },
  {
    "Containers": -1,
    "Created": 1505322880,
    "Id": "sha256:da5939581ac835614e3cf6c765e7489e6d0fc602a44e98c07013f1c938f49675",
    "Labels": {
      "maintainer": "NGINX Docker Maintainers <docker-maint@nginx.com>"
    },
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "nginx@sha256:aa1c5b5f864508ef5ad472c45c8d3b6ba34e5c0fb34aaea24acf4b0cee33187e"
    ],
    "RepoTags": [
      "nginx:latest"
    ],
    "SharedSize": -1,
    "Size": 154928640,
    "VirtualSize": 154928640
  },
  {
    "Containers": -1,
    "Created": 1505313146,
    "Id": "sha256:76da55c8019d7a47c347c0dceb7a6591144d232a7dd616242a367b8bed18ecbc",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "alpine@sha256:f006ecbb824d87947d0b51ab8488634bf69fe4094959d935c0c103f4820a417d"
    ],
    "RepoTags": [
      "alpine:latest"
    ],
    "SharedSize": -1,
    "Size": 6291968,
    "VirtualSize": 6291968
  },
  {
    "Containers": -1,
    "Created": 1505297656,
    "Id": "sha256:54511612f1c4d97e93430fc3d5dc2f05dfbe8fb7e6259b7351deeca95eaf2971",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "busybox@sha256:99ccecf3da28a93c063d5dddcdf69aeed44826d0db219aabc3d5178d47649dfa"
    ],
    "RepoTags": [
      "busybox:latest"
    ],
    "SharedSize": -1,
    "Size": 3408384,
    "VirtualSize": 3408384
  },
  {
    "Containers": -1,
    "Created": 1505255051,
    "Id": "sha256:05a3bd381fc2470695a35f230afefd7bf978b566253199c4ae5cc96fafa29b37",
    "Labels": null,
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "hello-world@sha256:1f19634d26995c320618d94e6f29c09c6589d5df3c063287a00e6de8458f8242"
    ],
    "RepoTags": [
      "hello-world:latest"
    ],
    "SharedSize": -1,
    "Size": 262656,
    "VirtualSize": 262656
  },
  {
    "Containers": -1,
    "Created": 1505107985,
    "Id": "sha256:98f8400d2944a44be7160f966ae8120f3b8c7096b01e45988e03aa8687c70eee",
    "Labels": {},
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "logstash@sha256:6d5236d5a2371af15d19300f80be7e742e4fa15a19335c6a1372e685e803bc70"
    ],
    "RepoTags": [
      "logstash:latest"
    ],
    "SharedSize": -1,
    "Size": 1098129408,
    "VirtualSize": 1098129408
  },
  {
    "Containers": -1,
    "Created": 1505099138,
    "Id": "sha256:d1ac13423d3cf7664e2b158049330c713f777473815b0ffe03d4103412505f0c",
    "Labels": {},
    "ParentId": "",
    "Platform": "linux/amd64",
    "RepoDigests": [
      "elasticsearch@sha256:3686a5757ed46c9dbcf00f6f71fce48ffc5413b193a80d1c46a21e7aad4c53ad"
    ],
    "RepoTags": [
      "elasticsearch:latest"
    ],
    "SharedSize": -1,
    "Size": 820256256,
    "VirtualSize": 820256256
  }
]
```

### CODE REVIEW

**Review by johnstep — APPROVED** [review]
LGTM

File: api/swagger.yaml
Comment (thaJeztah) [inline api/swagger.yaml]: Can you add;

- an `example:` key with a value (e.g. "linux/amd64")? 
- extend the description a bit to explain the field is omitted if empty, the information included (`<platform>/<architecture>`) (or is it "operating system"?)
- also update the examples section for the `/images/json` endpoint and add example values there? https://github.com/moby/moby/blob/master/api/swagger.yaml#L5965

Comment (thaJeztah) [reply api/swagger.yaml]: This also needs a line in https://github.com/moby/moby/blob/master/docs/api/version-history.md, but should go in a new API version; I prepared a PR for that here: https://github.com/moby/moby/pull/34908

That already has two LGTM's so we can merge if this PR gets ok'd

**Review by thaJeztah — CHANGES_REQUESTED** [review]
left some notes for the Swagger file / docs

**Comment by stevvooe** [comment]
So, the "specifier" string should only be used for user input. In this case, you want to report the platform with the OCI structure.

Also, images can have multiple platforms.

**Comment by jhowardmsft** [comment]
@stevvooe gotcha. Will update 😊

**Comment by jhowardmsft** [comment]
@thajeztah. Let me get the code first 😉

File: api/swagger.yaml
Comment (jhowardmsft) [reply api/swagger.yaml]: OK, so adding an example in the /images/json bit
With the latest update, the description bit goes away as it's now using the OCI image spec.

I'll follow up the version-history.md once that's merged.

**Comment by jhowardmsft** [comment]
OK, that's updated to use the OCI image spec. Unfortunately (for I suspect the same reasons auto-generation of the API .go files is turned off for much of the rest of the API), I had to disable it for image_summary.go. Added block comment in that file as to why. Here's what the API will provide now:


```
[
  {
    "Containers": -1,
    "Created": 1505777500,
    "Id": "sha256:2d696327ab2e15a3354ca258249358e244f3219eef76c653d0e6d3bd87f7830f",
    "Labels": null,
    "ParentId": "",
    "Platforms": [
      {
        "architecture": "amd64",
        "os": "linux"
      }
    ],
    "RepoDigests": [
      "ubuntu@sha256:31371c117d65387be2640b8254464102c36c4e23d2abe1f6f4667e47716483f1"
    ],
    "RepoTags": [
      "ubuntu:latest"
    ],
    "SharedSize": -1,
    "Size": 176949760,
    "VirtualSize": 176949760
  }
]
```

File: api/types/image_summary.go
Comment (thaJeztah) [inline api/types/image_summary.go]: Looks like there's `x-go-name` at least for the model; https://github.com/go-swagger/go-swagger/pull/713 but not sure if a similar thing exist for individual properties; @dnephin may know that

Comment (jhowardmsft) [reply api/types/image_summary.go]: Even if that works, it doesn't solve the `OSFeatures` to `os.features` problem. In reality, unless the version of go-swagger in use has a way to directly pull in an already defined struct (ie the Platform struct in the image-spec), we're kind of re-inventing the wheel...

Comment (thaJeztah) [reply api/types/image_summary.go]: Yup, understood that (was just curious if it was possible in general for other cases)

I actually was looking a while back at the option to reverse the direction, and generate the swagger from the code, but that would also mean that updating docs requires code-changes (e.g. if we want to have a nicer description for the docs)

Comment (dnephin) [reply api/types/image_summary.go]: `x-go-name` is supposed to work.

The source of truth should really be `api/types`. It's good that we mirror the `specs.Platform`, but if we use the imported struct the type could change without us really knowing about the change, and we could break API compatibility. 

It's also important to be consistent with other fields. The swagger generated type is a lot more consistent, so I would say don't use `x-go-name` at all. `os.features` is not like anything else we have in the API.

It's going we match the same fields, but I don't think we should use the imported struct.

Also why an array of pointers?

Comment (jhowardmsft) [reply api/types/image_summary.go]: > Also why an array of pointers?

See comment https://github.com/moby/moby/pull/34924#issuecomment-331308527 from @stevvooe 

@dnephin - I'm not quite sure what you're saying in the rest of it. Are you saying hard-code something that's identical to the OCI image spec `Platform`, but just not use the import?

File: api/types/image_summary.go
Comment (DHowett-MSFT) [inline api/types/image_summary.go]: nit: `meaningful` :smile:

File: api/types/image_summary.go
Comment (dnephin) [reply api/types/image_summary.go]: I don't see anything in that comment about pointers, could you elaborate?

I think it should have the same "form" as the OCI  Platform spec, but the field names should be consistent with the other fields in the API. So `OSFeatures` seems more appropriate than `os.features`.

Comment (stevvooe) [reply api/types/image_summary.go]: @dnephin We are following the OCI platform structure from image spec.

`os.features` was deprecated, so it actually shouldn't be defined here.

File: api/swagger.yaml
Comment (stevvooe) [inline api/swagger.yaml]: Features was also deprecated. Let's remove it.

File: api/swagger.yaml
Comment (stevvooe) [inline api/swagger.yaml]: `OSVersion` was deprecated, as well.

Comment (tianon) [reply api/swagger.yaml]: Wait, doesn't Windows use and rely on `os.version`? (I think the "relying on it" part was still a TODO, but was definitely planned, especially with the imminent RS3 that breaks compat)

Comment (friism) [reply api/swagger.yaml]: Yeah, +1

Comment (stevvooe) [reply api/swagger.yaml]: We asked for commentary and specification on this field from the MSFT side and ended up deprecating the field because we had no response. Perhaps, an annotation would be sufficient here?

Comment (tianon) [reply api/swagger.yaml]: See https://github.com/docker/distribution/issues/2398#issuecomment-333659799 for where it was confirmed that `os.version` is not in fact deprecated:

> Actually, it was left in place, apologies for the confusion.
> 
> The issue I see is that it is under-specified. What lands in that field? Do versions have an ordering? What is implied by that ordering? It would be good to have the microsoft folks work with OCI to define what is actually needed here, rather than have us guess.

**Comment by coolljt0725** [comment]
ping @jhowardmsft  what's the status of this pr?

**Comment by jhowardmsft** [comment]
@coolljt0725 It's still on my backlog. I'll try to get to it in the next few days.

**Comment by thaJeztah** [comment]
ping @jhowardmsft

**Comment by cpuguy83** [comment]
This seems inactive. Let's close for now and you can re-open when ready.
Hope that's ok!

Thanks! 🙇

**Comment by jhowardmsft** [comment]
We will need it for any form of LCOW CI. I'll open a follow-up if anyone ever cares about that. My preference would still be to get this in though, but there was far too much debate above and lack of agreement between maintainers.

**Comment by cpuguy83** [comment]
If we can get consensus that would be great.
Not sure I'm following where the issue is, though.

@stevvooe @tianon

**Pull request closed without merging**
